### DISCUSSION TITLE
Working Example of Bubble Chart - Hello World?

### DISCUSSION THREAD

**Discussion by MichaelCropper** [d3/d3#3650] [Q&A] [answered]
The docs on how to use D3 are abysmal to say the least. 

Is there a simple HelloWorld Bubble Chart example I can follow? 

Hello: 10
World: 20

That kind of data

Seems quite a few things have changed in recent versions of D3 but the content online is still often referring to v2, 3, 4 of the software which isn't that useful since those are 10 years old. 

And need an example that is just super simple. So none of this csv data that seems to be highly prevalent in all the docs for some unknown reason. No-one uses CSVs as a data source in 2023, this isn't the 1990s. 

Seems super powerful does this library, but honestly, can't even get a basic example working so it's really putting me off exploring this further if it is this difficult to use and understand the art of the possible.

**Comment by mbostock** [comment]
How’s this?

![untitled (59)](https://user-images.githubusercontent.com/230541/232895976-7ec603cb-7acc-493d-b20a-82794cedf57e.png)

```js
BubbleChart([["Hello", 10], ["World", 20]])
```

This snippet uses the [D3 bubble chart component](https://observablehq.com/@d3/bubble-chart) which you can easily import on Observable, you copy-paste the BubbleChart function into your web application. See [this example notebook](https://observablehq.com/d/4c617dfa997e6c95).

Putting your specific question aside, your tone is unproductive and unwelcome. I understand that you’re frustrated that the documentation isn’t giving you what you need, but please remember that D3 and its documentation are provided to you entirely for free. You get what you pay for. There’s no reason to insult us, the maintainers, by calling them “abysmal”—that only serves to discourage us and leads to maintainer burn out.

> No-one uses CSVs as a data source in 2023, this isn't the 1990s.

This is patently false. CSV is an incredibly popular plain text file format for tabular data, and has only become more popular in recent times. It is widely supported by software and simpler than many other alternatives (e.g., XSLX).

**Reply by MichaelCropper** [reply]
33% of the way there on the example - that single one liner, while incomplete, is actually more useful than anything I have read over the last week anywhere online. 

What is missing from that example is;
1) HTML <div> bits
2) <script> import bits

This is the kind of super basic example I'm looking for. I need a basic full working example that I can just copy and paste into my HelloWorld.html page that I'm editing with Microsoft Notepad so I can see it working in full. 

From what I can tell, the "Observable" website seems to be a kind of semi-SaaS (but free and with some funky git forking type tech) but all public. Personally I find it very confusing and all the features just aren't helpful and get in my way when I'm clicking about selecting text to try and copy and paste an example.

I've read hundreds of articles/pages/examples (and tried to implement way too many of them than I'd care to admit) and still not got a basic Bubble Chart working with the most recent version. 

I agree, my tone is not of joy, please don't take it personally. I don't want to come across as disgruntled, as you say, this is an open source project and I love open source and contribute to many projects in various ways. I'm just a little frustrated that a piece of technology that looks so awesome on all the examples I can see in the examples but can't take 10 mins to provide a basic full working documented example I can copy and paste to see working. It's like we've done 99% of the awesome work with the tech, but forgot to point the user where to put their key goes in their new car to start the engine so they can't do anything with it other than look at other examples in the show room and walk out empty and disappointed. 

Seriously @mbostock Mike, you seem to be the main guy behind this stuff, so extremely well done. The tech looks awesome. I just need to know how to use it :-)

Basic HelloWorld.html version. No package managers. No Observable. No CSVs. No 100x attribute configuration. No grouping/multi-layer hierarchy. Just an idiots guide to getting started. 

I need to be taken on a learning journey, a simple one. Just like with basic HTML. We all start with 

< html>
< head></head>
< body>
< h1>My First Website!</h1>
< /body>
< /html>

Right. Then we get deeper and deeper as we learn more about the next layers. The docs at present are throwing too much at me to absorb, to the point I can't get anything working. 

I don't even mind contributing to the project from a documentation perspective. If I can understand how the tech works I can write about it so it's easy for others to understand.

**Reply by mbostock** [reply, marked as answer]
Does this help?

```html
<!DOCTYPE html>
<body>
<script type="module">

import * as d3 from "https://cdn.jsdelivr.net/npm/d3@7/+esm";

const svg = BubbleChart([["Hello", 10], ["World", 20]]);
document.body.appendChild(svg);

function BubbleChart() {
  // … paste BubbleChart function here …
}

</script>
```

**Reply by curran** [reply]
@MichaelCropper For learning, you may like this video: [Constructing Visualizations: Marks and Channels](https://www.youtube.com/watch?v=48TUmsG_s6U&list=PL9yYRbwpkyks0RzTyTOLPYbHVSRJ68zOO&index=9). This shows basic usage of D3 to plot things on the screen, without any CSV. The rest of that [playlist](https://www.youtube.com/watch?v=mRNkZ3qHHBw&list=PL9yYRbwpkyks0RzTyTOLPYbHVSRJ68zOO&index=1) provides context for how to use D3 in general.

For example, here's the code that renders some circles, encapsulated as a function:

```js
export const xPosition = (selection, { data, height }) => {
  selection
    .selectAll('circle')
    .data(data)
    .join('circle')
    .attr('cx', (d) => d * 70 + 60)
    .attr('cy', height / 2)
    .attr('r', 30);
};
```
This is what that looks like:
![image](https://user-images.githubusercontent.com/68416/232969868-e2dbb6af-7a6e-4220-b35d-29b0b8ecc74d.png)

It's not a "bubble chart" or any chart really, but it's a simple precursor to a scatter plot that shows the structure of D3 rendering logic in its basic form. That can be expanded on to, for example, use color like this:

```js
const hue = (selection, { data, height }) => {
  selection
    .selectAll('circle')
    .data(data)
    .join('circle')
    .attr('cx', (d) => d * 70 + 60)
    .attr('cy', height / 2)
    .attr('r', 30)
    .attr('fill', (d, i) => d3.schemeCategory10[i]);
};
```

![image](https://user-images.githubusercontent.com/68416/232970261-86b31624-e2fb-439e-a82e-2412c0baba26.png)

The [video](https://www.youtube.com/watch?v=48TUmsG_s6U&list=PL9yYRbwpkyks0RzTyTOLPYbHVSRJ68zOO&index=9) explains in detail what each part of the code does and how these functions can be invoked. The invocation of this function in the example is somewhat convoluted, but a simple invocation would look like this:

```js
const svg = d3.select('body').append('svg');
xPosition(svg, {
  data: [0, 1, 2, 3, 4],
  height: 150
});
```

Code links:

 * [full working code as ES modules in VizHub](https://vizhub.com/curran/035e2a6bfb7e4486a7f82daff8850ad4?edit=files)
 * [full working code as vanilla JS in blocks.roadtolarissa.com](https://blocks.roadtolarissa.com/curran/6b5a43a958e7d9adea657a984ce1fbd6)

Hope this helps!

**Reply by MichaelCropper** [reply]
Thanks @mbostock - Yes that's exactly what is missing from the docs that I was looking for. 

For completeness, and for anyone else who comes across this post in the future. For Version 7 of D3, here is a copy and paste BubbleChart example. Simply put this into a HelloBubble.html file and it just works.....

(
Edit 1: Important note: please remove the space after ever < below, GitHub is stripping that info out annoyingly even though wrapped in a code block - so just remove these spaces when you copy and paste

Edit 2: GitHub stupid formatting still getting in the way of a copy and paste job. Updated code snippet so it hopefully now displays correctly and says what to do
)

`< !DOCTYPE html>
< html>
< head>< /head>
< body>
< script type="module">

import * as d3 from "https://cdn.jsdelivr.net/npm/d3@7/+esm";

const svg = BubbleChart([["Hello", 10], ["World", 20]]);
document.body.appendChild(svg);

// ***** Copy this whole function from here, https://observablehq.com/@d3/bubble-chart#BubbleChart, into below this line and it will work *****

< /script>
< /body>
< /html>`

**Comment by MichaelCropper** [comment]
So one thing that is confusing me @mbostock is this data structure;

[["Hello", 10], ["World", 20]]

This is a JavaScript multi-dimensional array. 

What about other more language agnostic data structures such as JSON? What would that look like? 

The only examples I can find online is reading from a abc.json type file locally, whereas in reality most use cases would be serializing a JSON object of some kind to a string dump to pass into the View layer of a traditional MVC type web app. 

I'm assuming that D3 supports what is essentially a basic array of key:value pairs in some way then we simply tell D3 what the Key Name is and hence can then loop through and get the Value bits. 

Just struggling to find a concrete example of what this would look like. 

Since most back end languages would struggle to translate a DB query with results in proper objects into this JavaScript centric syntax when dumping out. 

Is it possible to dump a JSON Array into that section instead? 

i.e. something along the lines of for an array of data containing NumberOfDaysInMonth for example;

`{
  "allData": [
    {
      "key": "January",
      "value": 31
    },
    {
      "key": "February",
      "value": 28
    }
  ]
}`

Then get D3 to essentially pluck the data such as with the pseudo code of;

`forEach {
    d3.allData.key
    and
    d3.allData.value
}`

**Comment by MichaelCropper** [comment]
Thanks @curran Feels like you created and published that 40 mins video just for me given it was just published at the time you responded! 

I'll find some time to have a watch of that video (likely multiple times!), had a quick skim through and seems like it's covering off a lot of the architectural / background / conceptual / important stuff that the main docs just seem to ignore. I'll no doubt have a lot of further questions after having a deeper play. 

Good news is though, I've got a basic version actually working so I can actually start to have a play with the tech that to date has just been teasing me. So hopefully I can figure out the icing on the cake from here (colours, animations, transitions etc.)

I have to say though, the small snippets I have watched in the video, I really like you're style. It assumes I'm an idiot which is exactly what is needed. It assumes I know nothing (and I do know nothing about D3!). It helps me understand and guide me.

**Comment by MichaelCropper** [comment]
Probably a separate question, but what's this website https://d3-graph-gallery.com/graph/bubble_basic.html ? 

I'm just confused as I thought the main website for info is https://d3js.org/ ?

But then I also came across https://github.com/d3/d3/blob/main/API.md which seems to be API documentation?

And to add to confusion I also came across https://devdocs.io/d3~7/

Really confusing. I just don't know what to trust for an official source of information. 

Any insights into what I should be trusting / ignoring? Where is the official source of truth from a documentation perspective?
